src/Orleans.Core/Providers/LegacyProviderConfigurator.cs

```diff
@@ -0,0 +1,168 @@
+                {
+                    services.AddSingleton<ILogConsistencyProvider>(sp => sp.GetServiceByName<ILogConsistencyProvider>(ProviderConstants.DEFAULT_LOG_CONSISTENCY_PROVIDER_NAME));
+                    foreach (IProviderConfiguration providerConfig in providerGroup.SelectMany(kvp => kvp.Value.Providers.Values))
+                    {
+                        RegisterProvider<ILogConsistencyProvider>(providerConfig, services, defaultInitStage, defaultStartStage);
+                    }
+                }
+                else if (providerGroup.Key == ProviderCategoryConfiguration.STATISTICS_PROVIDER_CATEGORY_NAME)
+                {
+                    IProviderConfiguration providerConfig = providerGroup.SelectMany(kvp => kvp.Value.Providers.Values).FirstOrDefault();
+                    if(providerConfig != null)
+                    {
+                        // Looks like we only support a single statistics provider that can be any of the publisher interfaces. fml
+                        // TODO: Kill our statistics system.. please!?
+                        services.AddSingleton<IStatisticsPublisher>(sp => sp.GetServiceByName<IProvider>(providerConfig.Name) as IStatisticsPublisher);
+                        services.AddSingleton<ISiloMetricsDataPublisher>(sp => sp.GetServiceByName<IProvider>(providerConfig.Name) as ISiloMetricsDataPublisher);
+                        services.AddSingleton<IClientMetricsDataPublisher>(sp => sp.GetServiceByName<IProvider>(providerConfig.Name) as IClientMetricsDataPublisher);
+                        RegisterProvider<IProvider>(providerConfig, services, defaultInitStage, defaultStartStage);
+                    }
+                }
+                else
+                {
+                    foreach (IProviderConfiguration providerConfig in providerGroup.SelectMany(kvp => kvp.Value.Providers.Values))
+                    {
+                        RegisterProvider<IProvider>(providerConfig, services, defaultInitStage, defaultStartStage);
+                    }
+                }
+            }
+        }
+
+        private class ProviderLifecycleParticipant<TService> : ILifecycleParticipant<TLifeCycle>
+            where TService : class
+        {
+            private readonly ILogger logger;
+            private readonly IProviderConfiguration config;
+            private readonly IServiceProvider services;
+            private readonly LegacyProviderConfigurator.ScheduleTask schedule;
+            private readonly int defaultInitStage;
+            private int initStage;
+            private readonly int defaultStartStage;
+            private int startStage;
+
+            public ProviderLifecycleParticipant(IProviderConfiguration config, IServiceProvider services, ILoggerFactory loggerFactory, int defaultInitStage, int defaultStartStage)
+            {
+                this.logger = loggerFactory.CreateLogger(config.Type);
```

Comment:
This Type string is free-form, so it might look ugly and non-uniform in the logs - possibly `namespace.type, assembly`? Maybe including the strong name of the assembly?

Reply:
All we have at this point is the "Type" string provided by the applications configuration.  We can't even convert it to an actual type at this point.  Open to other suggestions but our choices are somewhat limited..

Reply:
Is it possible that we can resolve the string to a real type earlier on?

Reply:
maybe.  Let me look into that.  Initially I thought we needed to wait until the assembly loading (parts) had occurred, so I delayed it as late as possible, but I'm less convinced of that now.  Will take a look.  Resolving the actual types here would make this code cleaner.